api/src/main/java/jakarta/mail/Session.java

```diff
@@ -1102,7 +1133,8 @@ private void loadProvidersFromStream(InputStream is) throws IOException {
      */
     public synchronized void addProvider(Provider provider) {
         providers.add(provider);
-        providersByClassName.put(provider.getClassName(), provider);
+        if (!providersByClassName.containsKey(provider.getClassName()))
```

Comment:
I was confused here. This is now necessary to avoid jakarta resource overwrite javamail. But why was not this added here before, if we were doing this check in `providersByProtocol` already?.